os/drivers/video/video_null_driver.c

```diff
@@ -0,0 +1,1554 @@
+ * Private Function Prototypes
+ ****************************************************************************/
+static int video_null_open(FAR void *video_priv);
+static int video_null_close(FAR void *video_priv);
+static int video_null_do_halfpush(FAR void *video_priv, bool enable);
+static int video_null_set_buftype(FAR void *video_priv, enum v4l2_buf_type type);
+static int video_null_set_buf(FAR void *video_priv, uint32_t bufaddr, uint32_t bufsize);
+static int video_null_cancel_dma(FAR void *video_priv);
+static int video_null_get_range_of_fmt(FAR void *video_priv, FAR struct v4l2_fmtdesc *format);
+static int video_null_get_range_of_framesize(FAR void *video_priv, FAR struct v4l2_frmsizeenum *frmsize);
+static int video_null_try_format(FAR void *video_priv, FAR struct v4l2_format *format);
+static int video_null_set_format(FAR void *video_priv, FAR struct v4l2_format *format);
+static int video_null_get_range_of_frameinterval(FAR void *video_priv, FAR struct v4l2_frmivalenum *frmival);
+static int video_null_set_frameinterval(FAR void *video_priv, FAR struct v4l2_streamparm *parm);
+static int video_null_get_range_of_ctrlvalue(FAR void *video_priv, FAR struct v4l2_query_ext_ctrl *range);
+static int video_null_get_menu_of_ctrlvalue(FAR void *video_priv, FAR struct v4l2_querymenu *menu);
+static int video_null_get_ctrlvalue(FAR void *video_priv, uint16_t ctrl_class, FAR struct v4l2_ext_control *control);
+static int video_null_set_ctrlvalue(FAR void *video_priv, uint16_t ctrl_class, FAR struct v4l2_ext_control *control);
+static int video_null_refresh(FAR void *video_priv);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+struct v4l2_mapping_s {
+	uint16_t ctrl_class;		/* Control class */
+	uint16_t id;				/* Control id */
+	uint16_t type;				/* enum #v4l2_ctrl_type */
+	char name[32];				/* Name of control */
+	int32_t min;				/* Minimum value */
+	int32_t max;				/* Maximum value */
+	uint32_t step;				/* step */
+	int32_t default_value;		/* Default value */
+	int32_t current_value;		/* Current value */
+};
+
+struct v4l2_interval_s {
+	uint32_t min_numerator;		/* numerator */
+	uint32_t max_numerator;		/* numerator */
+	uint32_t step_numerator;	/* numerator */
+};
+
+struct v4l2_frames_s {
+	uint32_t wWidth;
+	uint32_t wHeight;
+	uint32_t type;
+	struct v4l2_interval_s intervals[MAX_INTERVAL];
+	uint32_t intervalcount;
+};
+
+struct v4l2_format_s {
+	char description[V4L2_FMT_DSC_MAX];	/* Description string */
+	uint16_t type;				/* v4l2_buf_type */
+	uint32_t pixelformat;		/* Format fourcc      */
+	uint32_t colorspace;		/* Color space */
+	uint8_t bpp;				/* Bits per pixel */
+	struct v4l2_frames_s frames[MAX_FRAMES];	/* Pointer to frames object */
+	uint16_t framecount;		/* Number of frame supported */
+};
+
+struct dummy_null_priv_s {
+	int16_t crefs;				/* Reference count on the driver instance */
```

Comment:
continue comment in line94, many things in below can be added in struct video_lowerhalf_s as a common lower's functionality.

Reply:
To make null_video driver non-dependent on video driver, we declared separately

Reply:
@Taejun-Kwon: Yes we made null_driver as low level driver only, below video layer.